trace/trace.go

```diff
@@ -0,0 +1,228 @@
+	defer span.End()
+	return t.ds.Get(ctx, key)
+}
+
+// Has implements the ds.Datastore interface.
+func (t *Datastore) Has(ctx context.Context, key ds.Key) (exists bool, err error) {
+	ctx, span := t.tracer.Start(ctx, "Has", otel.WithAttributes(attribute.String("key", key.String())))
+	defer span.End()
+	return t.ds.Has(ctx, key)
+}
+
+// GetSize implements the ds.Datastore interface.
+func (t *Datastore) GetSize(ctx context.Context, key ds.Key) (size int, err error) {
+	ctx, span := t.tracer.Start(ctx, "GetSize", otel.WithAttributes(attribute.String("key", key.String())))
+	defer span.End()
+	return t.ds.GetSize(ctx, key)
+}
+
+// Delete implements the ds.Datastore interface.
+func (t *Datastore) Delete(ctx context.Context, key ds.Key) (err error) {
+	ctx, span := t.tracer.Start(ctx, "Delete", otel.WithAttributes(attribute.String("key", key.String())))
+	defer span.End()
+	return t.ds.Delete(ctx, key)
+}
+
+// Query implements the ds.Datastore interface.
+func (t *Datastore) Query(ctx context.Context, q dsq.Query) (dsq.Results, error) {
+	ctx, span := t.tracer.Start(ctx, "Query", otel.WithAttributes(attribute.String("query", q.String())))
+	defer span.End()
+	return t.ds.Query(ctx, q)
+}
+
+// Batch implements the ds.Batching interface.
+func (t *Datastore) Batch(ctx context.Context) (ds.Batch, error) {
+	ctx, span := t.tracer.Start(ctx, "Batch")
+	defer span.End()
+
+	if batch, ok := t.ds.(ds.Batching); ok {
+		return batch.Batch(ctx)
+	}
+
+	return ds.NewBasicBatch(t), nil
+}
+
+// DiskUsage implements the ds.PersistentDatastore interface.
+func (t *Datastore) DiskUsage(ctx context.Context) (uint64, error) {
+	ctx, span := t.tracer.Start(ctx, "DiskUsage")
+	defer span.End()
+	return ds.DiskUsage(ctx, t.ds)
+}
+
+// Scrub implements the ds.ScrubbedDatastore interface.
+func (t *Datastore) Scrub(ctx context.Context) error {
+	ctx, span := t.tracer.Start(ctx, "Scrub")
+	defer span.End()
+
+	if dstore, ok := t.tracer.(ds.ScrubbedDatastore); ok {
+		return dstore.Scrub(ctx)
+	}
+
+	return nil
+}
+
+// CollectGarbage implements the ds.GCDatastore interface.
+func (t *Datastore) CollectGarbage(ctx context.Context) error {
+	ctx, span := t.tracer.Start(ctx, "CollectGarbage")
+	defer span.End()
+
+	if dstore, ok := t.tracer.(ds.GCDatastore); ok {
+		return dstore.CollectGarbage(ctx)
+	}
+
+	return nil
+}
+
+// Check implements the ds.CheckedDatastore interface.
+func (t *Datastore) Check(ctx context.Context) error {
+	ctx, span := t.tracer.Start(ctx, "Check")
+	defer span.End()
+
+	if dstore, ok := t.tracer.(ds.CheckedDatastore); ok {
+		return dstore.Check(ctx)
+	}
+
+	return nil
+}
+
+// NewTransaction implements the ds.TxnDatastore interface.
+func (t *Datastore) NewTransaction(ctx context.Context, readOnly bool) (ds.Txn, error) {
+	ctx, span := t.tracer.Start(ctx, "NewTransaction", otel.WithAttributes(attribute.Bool("readOnly", readOnly)))
+	defer span.End()
```

Comment:
Is there any way to associate gets/puts within the transaction with the transaction's span?

Reply:
I wouldn't know how to associate that :/ 

I guess the association would happen at the level where the transaction is created. We could carry around the span context in the `Txn` struct but then what would you do with the context that gets passed into transaction put/gets?

Reply:
Yeah, fair enough. I'd just leave it as-is.